canvas/canvas_course_usage.py

```diff
@@ -0,0 +1,107 @@
+import logging
+import time
+import pandas as pd
+from json.decoder import JSONDecodeError
+from requests_futures.sessions import FuturesSession
+from concurrent.futures import as_completed
+import json
+logger = logging.getLogger(__name__)
+
+
+class CanvasCourseUsage:
+    def __init__(self, canvas_url, canvas_token, retry_attempts, course_ids):
+        self.canvas_url = canvas_url,
+        self.canvas_token = canvas_token,
+        self.course_ids = course_ids
+        self.retry_attempts = retry_attempts
+        self.canvas_usage_courses = []
+        self.course_retry_list = []
+        self.retry_count = 0
+
+    def parsing_canvas_course_usage_data(self, response):
+        logger.info("parsing_canvas_course_usage_data Call")
+        new_dic = {}
+        if response is None:
+            logger.info(f"For Canvas Course usage response is None ")
+            return
+
+        logger.info(f"CanvasCourseUsage date collected so far : {len(self.canvas_usage_courses)}")
+        status = response.result().status_code
+
+        course_id = response.result().url.split('courses/')[1].split('/')[0]
+
+        status_codes_to_retry = [500]
+        if status != 200:
+            logger.info(f'retry_count: {self.retry_count}')
+            if self.retry_count < self.retry_attempts and status in status_codes_to_retry:
+                logger.info("Append to retry list")
+                self.course_retry_list.append(course_id)
+            logger.info(f"Response not successful with status code {status} due to {response.result().text}")
+            return
+
+        try:
+            analytics_data = json.loads(response.result().text)
+        except JSONDecodeError as e:
+            logger.error(f"Error in parsing the response due to {e.msg}")
+            return
```

Comment:
I think you could also put this in the re-try mechanism, right? Maybe use a flag, something like what I did in my async PR with `problem_encountered`? I think at some point we may want to refactor so this mechanism is the same in all the classes we've written.

Reply:
https://github.com/tl-its-umich-edu/course-inventory/pull/51#issuecomment-608623356

Reply:
Please see my reply below, so you understand why I'm making this point. I do think you probably want to re-try in these cases. If it fails after three tries, it'll move on.